iroh-bytes/src/downloader.rs

```diff
@@ -454,540 +503,495 @@ struct Service<G: Getter, D: Dialer> {
     fn new(
+        db: DB,
         getter: G,
         dialer: D,
         concurrency_limits: ConcurrencyLimits,
         msg_rx: mpsc::Receiver<Message>,
     ) -> Self {
         Service {
             getter,
-            providers: ProviderMap::default(),
             dialer,
-            concurrency_limits,
             msg_rx,
-            nodes: HashMap::default(),
+            concurrency_limits,
+            nodes: Default::default(),
+            providers: Default::default(),
+            requests: Default::default(),
             goodbye_nodes_queue: delay_queue::DelayQueue::default(),
-            current_requests: HashMap::default(),
+            active_requests: Default::default(),
             in_progress_downloads: Default::default(),
-            scheduled_requests: HashMap::default(),
-            scheduled_request_queue: delay_queue::DelayQueue::default(),
+            progress_tracker: ProgressTracker::new(),
+            intents: Default::default(),
+            queue: Default::default(),
+            db,
         }
     }
 
     /// Main loop for the service.
     async fn run(mut self) {
         loop {
-            // check if we have capacity to dequeue another scheduled request
-            let at_capacity = self
-                .concurrency_limits
-                .at_requests_capacity(self.in_progress_downloads.len());
-
+            trace!("wait for tick");
             tokio::select! {
                 Some((node, conn_result)) = self.dialer.next() => {
-                    trace!("tick: connection ready");
+                    trace!(node=%node.fmt_short(), "tick: connection ready");
                     self.on_connection_ready(node, conn_result);
                 }
                 maybe_msg = self.msg_rx.recv() => {
                     trace!(msg=?maybe_msg, "tick: message received");
                     match maybe_msg {
-                        Some(msg) => self.handle_message(msg),
+                        Some(msg) => self.handle_message(msg).await,
                         None => return self.shutdown().await,
                     }
                 }
-                Some(res) = self.in_progress_downloads.join_next() => {
+                Some(res) = self.in_progress_downloads.join_next(), if !self.in_progress_downloads.is_empty() => {
                     match res {
                         Ok((kind, result)) => {
-                            trace!("tick: download completed");
-                            self.on_download_completed(kind, result);
+                            trace!(kind=%kind.fmt_short(), "tick: transfer completed");
+                            self.on_download_completed(kind, result).await;
                         }
-                        Err(e) => {
-                            warn!("download issue: {:?}", e);
+                        Err(err) => {
+                            warn!(?err, "transfer task paniced");
```

Comment:
```suggestion
                            warn!(?err, "transfer task panicked");
```

However, it seems to be a `JoinError` can be both a cancel and panic

Reply:
:+1: We don't cancel the tasks in here though, we only use a cancellation token so I think panic is the only reason this can get triggered.